src/test/java/io/nats/client/impl/ErrorListenerTests.java

```diff
@@ -231,4 +232,31 @@ public void testExceptionInExceptionHandler() throws Exception {
             }
         }
     }
+
+    @Test
+    public void testDiscardedMessage() throws Exception {
```

Comment:
Can you update to shut down the test server after connect, so the outgoing queue is clearly being used, rather than having it work because of timing or the test server not actually reading.

Reply:
Hi Stephen, thanks for the quick review. I was able to confirm that this test case is indeterministic.

I discovered a corner case on flushing the connection. The flush command pushes a ping message on the queue. If the queue is full, this ping message may trigger an IllegalStateException (existing behaviour, when it cannot be offered within 5 seconds), or may be discarded (this pull request, when it cannot be offered directly). I will look into keeping the old behaviour for internal messages.

Reply:
Hi Stephen,

I have changed the implementation to make sure that upon a full outgoing queue:
* internal messages wait for the queue to become available
* messages from fast producers are discarded immediately 

Tests have been added.